File: day1/windturbines/submissions/accepted/jb.cc
#include <bits/stdc++.h>
using namespace std;

#define rep(i, a, b) for(int i = a; i < (b); ++i)
#define all(x) begin(x), end(x)
#define sz(x) (int)(x).size()
typedef long long ll;
typedef vector<int> vi;

struct UF {
  vi e;
  UF(int n) : e(n, -1) {}
  bool sameSet(int a, int b) { return find(a) == find(b); }
  int size(int x) { return -e[find(x)]; }
  int find(int x) { return e[x] < 0 ? x : e[x] = find(e[x]); }
  bool join(int a, int b) {
    a = find(a), b = find(b);
    if (a == b) return false;
    if (e[a] > e[b]) swap(a, b);
    e[a] += e[b]; e[b] = a;
    return true;
  }
};

struct FT {
  vector<ll> s;
  FT(int n) : s(n) {}
  void update(int pos, ll dif) { // a[pos] += dif
    for (; pos < sz(s); pos |= pos + 1) s[pos] += dif;
  }
  ll query(int pos) { // sum of values in [0, pos)
    ll res = 0;
    for (; pos > 0; pos &= pos - 1) res += s[pos-1];
    return res;
  }
};

template<class T>
struct RMQ {
  vector<vector<T>> jmp;
  RMQ(const vector<T>& V) : jmp(1, V) {
    for (int pw = 1, k = 1; pw * 2 <= sz(V); pw *= 2, ++k) {
      jmp.emplace_back(sz(V) - pw * 2 + 1);
      rep(j,0,sz(jmp[k]))
        jmp[k][j] = min(jmp[k - 1][j], jmp[k - 1][j + pw]);
    }
  }
  T query(int a, int b) {
    //assert(a < b); // or return inf if a == b
    int dep = 31 - __builtin_clz(b - a);
    return min(jmp[dep][a], jmp[dep][b - (1 << dep)]);
  }
};

struct LCA {
  int T = 0;
  vi time, path, ret;
  RMQ<int> rmq;

  LCA(vector<vi>& C) : time(sz(C)), rmq((dfs(C,sz(C)-1,-1), ret)) {}
  void dfs(vector<vi>& C, int v, int par) {
    time[v] = T++;
    for (int y : C[v]) if (y != par) {
      path.push_back(v), ret.push_back(time[v]);
      dfs(C, y, v);
    }
  }

  int lca(int a, int b) {
    if (a == b) return a;
    tie(a, b) = minmax({time[a], time[b]});
    return path[rmq.query(a, b)];
  }
};

int main() {
  cin.tie(0)->sync_with_stdio(0);
  cin.exceptions(cin.failbit);
  int n, m, q;
  cin >> n >> m >> q;
  vector<tuple<ll,int,int>> edges;
  rep(i,0,m) {
    int c, x, y;
    cin>>x>>y>>c;
    edges.emplace_back(c,x,y);
  }
  rep(i,1,n) edges.emplace_back(1e12,i-1,i);
  sort(all(edges));

  UF uf(n);
  vi root(n);
  vector<ll> weight(n);
  iota(all(root),0);
  vector<vi> childs(n);

  ll cost = 0;
  for(auto [w,a,b] : edges) {
    a = uf.find(a);
    b = uf.find(b);
    if(!uf.join(a,b)) continue;
    cost += w;
    int nxt = sz(childs);
    childs.emplace_back();
    childs.back() = {root[a],root[b]};
    weight.emplace_back(w);
    root[a] = root[b] = nxt;
  }

  vi leaf_chain(sz(childs),n);
  LCA lca(childs);
  FT fw(n+1);

  auto activate = [&](int x) {
    int r = sz(childs)-1;
    while(leaf_chain[r] != n) {
      int y = leaf_chain[r];
      int z = lca.lca(x, y);
      //assert(sz(childs[z]) == 2);
      int zx = childs[z][0], zy = childs[z][1];
      if(lca.lca(x,zx) != zx) swap(zx,zy);
      //assert(lca.lca(x,zx) == zx && lca.lca(y,zy) == zy);
      fw.update(leaf_chain[zx],-weight[z]);
      fw.update(y,weight[z]);
      leaf_chain[zy] = y;
      leaf_chain[r] = n;
      r = zx;
    }
    leaf_chain.back() = x;
  };

  vi ql(q), qr(q);
  vector<ll> ans(q);
  vector<vi> queries_at_l(n);
  rep(i,0,q) {
    cin>>ql[i]>>qr[i];
    ++qr[i];
    queries_at_l[ql[i]].emplace_back(i);
  }

  for(int l = n; l--;) {
    activate(l);
    for(auto i : queries_at_l[l])
      ans[i] = cost - fw.query(qr[i]);
  }

  for(auto a : ans)
    cout << a   << "\n";
}

File: day1/windturbines/submissions/accepted/jb.py
#!/usr/bin/env python3

# UF, FenwickTree, RMQ, and LCA from https://github.com/cheran-senthil/PyRival
class UnionFind:
  def __init__(self, n):
    self.parent = list(range(n))

  def find(self, a):
    acopy = a
    while a != self.parent[a]:
      a = self.parent[a]
    while acopy != a:
      self.parent[acopy], acopy = a, self.parent[acopy]
    return a

  def union(self, a, b): # must have self.find(b) = b
    self.parent[b] = a

class FenwickTree:
  def __init__(self, n):
    self.bit = [0]*n
    self.n = n

  def update(self, idx, x):
    while idx < self.n:
      self.bit[idx] += x
      idx |= idx + 1

  def query(self, end):
    x = 0
    while end:
      x += self.bit[end - 1]
      end &= end - 1
    return x

class RangeQuery:
  def __init__(self, data):
    self._data = _data = [list(data)]
    i, n = 1, len(_data[0])
    while 2 * i <= n:
      prev = _data[-1]
      _data.append([min(prev[j], prev[j + i]) for j in range(n - 2 * i + 1)])
      i <<= 1

  def query(self, begin, end):
    depth = (end - begin).bit_length() - 1
    return min(self._data[depth][begin], self._data[depth][end - (1 << depth)])

class LCA:
  def __init__(self, root, graph):
    self.time = [-1] * len(graph)
    self.path = [-1] * len(graph)
    P = [-1] * len(graph)
    t = -1
    dfs = [root]
    while dfs:
      node = dfs.pop()
      self.path[t] = P[node]
      self.time[node] = t = t + 1
      for nei in graph[node]:
        P[nei] = node
        dfs.append(nei)
    self.rmq = RangeQuery(self.time[node] for node in self.path)

  def __call__(self, a, b):
    if a == b:
        return a
    a = self.time[a]
    b = self.time[b]
    if a > b:
        a, b = b, a
    return self.path[self.rmq.query(a, b)]

##### main solution begins here #####

n, m, q  = [int(x) for x in input().split()]
edges = [(i,i+1,int(1e12)) for i in range(n-1)]
for i in range(m):
  edges.append([int(x) for x in input().split()])
edges.sort(key=lambda x: x[2])

uf = UnionFind(n)
root = [i for i in range(n)]
weight = [0]*(2*n-1)
childs = [[] for i in range(2*n-1)]
rt = n-1

for x, y, w in edges:
  x = uf.find(x)
  y = uf.find(y)
  if x == y:
    continue
  uf.union(x,y)
  rt += 1
  childs[rt] = [root[x], root[y]]
  weight[rt] = w
  root[x] = root[y] = rt

cost = sum(weight)

leaf_chain = [n]*len(childs)
lca = LCA(rt, childs)
fw = FenwickTree(n)

def activate(x):
  r = rt
  while leaf_chain[r] != n:
    y = leaf_chain[r]
    z = lca(x,y)
    zy, zx = childs[z]
    if lca.time[x] >= lca.time[zy]:
      zx, zy = zy, zx
    fw.update(leaf_chain[zx],-weight[z])
    fw.update(y,weight[z])
    leaf_chain[zy] = y
    leaf_chain[r] = n
    r = zx
  leaf_chain[-1] = x

qr, ans = [0]*q, [0]*q
queries_at_l = [[] for i in range(n)]

for i in range(q):
  ql, qr[i] = [int(i) for i in input().split()]
  queries_at_l[ql].append(i)

for l in reversed(range(n)):
  activate(l)
  for i in queries_at_l[l]:
    ans[i] = cost - fw.query(qr[i]+1)

print(*[a for a in ans])

